osbuild: new stage 'cacert' (HMS-4839)

lzap opened the pull request:
This pull request adds a new stage called 'cacert' to the osbuild package. It also includes file changes to support the CACustomization feature, which allows users to specify a list of certificates for the CA (Certificate Authority). The changes include updates to the Customizations struct, the osCustomizations function, and the OSCustomizations struct. Additionally, a new function called parseCerts is added to parse the certificate strings. The changes also include updates to the serialize and prependKernelCmdlineStage functions in the OS struct. Finally, a new file called ca_stage.go is added to the osbuild package, which contains the implementation of the NewCAStageStage function.

Needs: https://github.com/osbuild/osbuild/pull/1854

lzap commented on pkg/manifest/os.go:
```diff
@@ -863,9 +868,51 @@ func (p *OS) serialize() osbuild.Pipeline {
+	if len(p.CACerts) > 0 {
+		for _, cc := range p.CACerts {
+			for _, c := range parseCerts(cc) {
+				path := filepath.Join("/etc/pki/ca-trust/source/anchors", filepath.Base(c.SerialNumber.Text(16)))
```
Missing the filename extension, will add during review.

achilleas-k reviewed (changes requested):
Can we break down this PR into multiple commits for easier reviewing and history?

I like to have commits that introduce new things separate from the following ones that use them, so maybe:
1. Add `osbuild/..._stage.go`
2. Add `manifest/os.go` changes.
3. Add blueprint customization.
4. Add config option to `all-customizations.json`.

(though, 2 and 3 can be merged).

lzap commented on pkg/blueprint/customizations.go:
```diff
@@ -32,6 +32,7 @@ type Customizations struct {
 	Installer          *InstallerCustomization        `json:"installer,omitempty" toml:"installer,omitempty"`
 	RPM                *RPMCustomization              `json:"rpm,omitempty" toml:"rpm,omitempty"`
 	RHSM               *RHSMCustomization             `json:"rhsm,omitempty" toml:"rhsm,omitempty"`
+	CA                 *CACustomization               `json:"ca,omitempty" toml:"ca,omitempty"`
```
Sure, in the same struct I see a two-worded JSON field SSHKey as `sshkey` so I will stick with this theme and go for `cacerts`.

lzap commented:
Sure, amended all remarks. If you don’t mind, I am keeping one commit for easier work from my side, multiple commits rebasing is a pain to work with this isn’t a big PR by any means. I will split it after the code is ready.

mvo5 commented on pkg/manifest/os.go:
```diff
@@ -863,9 +868,51 @@ func (p *OS) serialize() osbuild.Pipeline {
+	for _, block := range blocks {
+		cert, err := x509.ParseCertificate(block)
+		if err != nil {
+			logrus.Warnf("failed to parse certificate, skipping: %s", err)
```
Is this the behavior we want/expect? This will mean that pasing an invalid cert string will just result in an empty list and the cert not appearing in the manifest. As a user I would probably be surprised if I paste a cert with an error and then get an image (much later) that does not contain it. Sorry, I think this is a very wordy way of saysing: "it seems to me we should return an error here" (unless I miss something of course which is entriely possible)

lzap replied:
Good call, explicit error it is.

mvo5 commented on pkg/manifest/os.go:
```diff
@@ -22,6 +24,7 @@ import (
 	"github.com/osbuild/images/pkg/platform"
 	"github.com/osbuild/images/pkg/rhsm/facts"
 	"github.com/osbuild/images/pkg/rpmmd"
+	"github.com/sirupsen/logrus"
```
(nipick) in general our convention for imports is (in most go project btw):
- stdlib
- external third party
- our own includes

we are not very strict about it but here I would suggest to move it between "strings" and github.com/osbuild/images/internal/common" (with \n before/after)

mvo5 commented on pkg/blueprint/customizations.go:
```diff
@@ -439,3 +440,10 @@ func (c *Customizations) GetRHSM() *RHSMCustomization {
+
+func (c *Customizations) GetCACerts() *CACustomization {
+	if c == nil {
+		return nil
```
I wonder if it wouldn't be better to also validate here, this would avoid the panic for invalid certs later. Here we could follow the pattern from e.g. GetRepositories() and return (*CACustomization,error) and then the callers (in fedora for example) can report a meaningful error for invalid cert strings early. Wdyt?

lzap replied:
I can do that, just note this introduces a dependency on distro/manifest package.

lzap replied:
Ah it is an import cycle actually, I cannot do that, do you want me to do it? I would suggest to create an explicit `CheckCACerts` function and extract the parsing X509 code into a separate package.

mvo5 replied:
Thanks for looking into this, I'm not sure I follow, maybe I'm missing something but I was thinking of something like this:
```go
diff --git a/pkg/blueprint/customizations.go b/pkg/blueprint/customizations.go
index 0bb5bacbe..f0d800077 100644
--- a/pkg/blueprint/customizations.go
+++ b/pkg/blueprint/customizations.go
@@ -428,12 +428,12 @@ func (c *Customizations) GetRHSM() *RHSMCustomization {
        return c.RHSM
 }
 
-func (c *Customizations) CheckCACerts() error {
+func (c *Customizations) checkCACerts() error {
        if c == nil {
                return nil
        }
 
-       if c.CACerts == nil {
+       if c.CACerts != nil {
                for _, bundle := range c.CACerts.PEMCerts {
                        _, err := cert.ParseCerts(bundle)
                        if err != nil {
@@ -445,9 +445,14 @@ func (c *Customizations) CheckCACerts() error {
        return nil
 }
 
-func (c *Customizations) GetCACerts() *CACustomization {
+func (c *Customizations) GetCACerts() (*CACustomization, error) {
        if c == nil {
-               return nil
+               return nil, nil
        }
-       return c.CACerts
+
+       if err := c.checkCACerts(); err != nil {
+               return nil, err
+       }
+
+       return c.CACerts, nil
 }
diff --git a/pkg/distro/fedora/images.go b/pkg/distro/fedora/images.go
index b0e737468..4c04a06c4 100644
--- a/pkg/distro/fedora/images.go
+++ b/pkg/distro/fedora/images.go
@@ -226,10 +226,11 @@ func osCustomizations(
        osc.Files = append(osc.Files, imageConfig.Files...)
        osc.Directories = append(osc.Directories, imageConfig.Directories...)
 
-       if err := c.CheckCACerts(); err != nil {
+       ca, err := c.GetCACerts()
+       if err != nil {
                panic(fmt.Sprintf("unexpected error checking CA certs: %v", err))
        }
-       if ca := c.GetCACerts(); ca != nil {
+       if ca != nil {
                osc.CACerts = ca.PEMCerts
        }
 
diff --git a/pkg/distro/rhel/images.go b/pkg/distro/rhel/images.go
index 5e13afe7f..806463b92 100644
--- a/pkg/distro/rhel/images.go
+++ b/pkg/distro/rhel/images.go
@@ -271,10 +271,11 @@ func osCustomizations(
                osc.NoBLS = *imageConfig.NoBLS
        }
 
-       if err := c.CheckCACerts(); err != nil {
+       ca, err := c.GetCACerts()
+       if err != nil {
                panic(fmt.Sprintf("unexpected error checking CA certs: %v", err))
        }
-       if ca := c.GetCACerts(); ca != nil {
+       if ca != nil {
                osc.CACerts = ca.PEMCerts
        }
 
```
i.e. very time we call GetCACerts() we implicitly check as well - this way we will not forget and it's similar to the pattern of GetRepositories() ?

mvo5 replied:
This is added now, marking it as resolved (and THANK YOU)

mvo5 commented on pkg/manifest/os.go:
```diff
@@ -863,9 +868,51 @@ func (p *OS) serialize() osbuild.Pipeline {
 	return pipeline
 }
 
+func parseCerts(cert string) []*x509.Certificate {
```
A simple unit test for this would be very nice, ideally both for a "positive" and "error" case, something like
```go
// taken from osbuild:test/data/certs/cert1.pem
const exampleCert = `
-----BEGIN CERTIFICATE-----
MIIDhTCCAm2gAwIBAgIUVya7VJ3O8W8SqwuEa0BZ4HSsXvAwDQYJKoZIhvcNAQEL
BQAwUTELMAkGA1UEBhMCREUxDzANBgNVBAgMBkJlcmxpbjEPMA0GA1UEBwwGQmVy
bGluMQwwCgYDVQQKDANPcmcxEjAQBgNVBAMMCWxvY2FsaG9zdDAgFw0yNDA4MjYx
MDQyNDBaGA8yMTI0MDgwMjEwNDI0MFowUTELMAkGA1UEBhMCREUxDzANBgNVBAgM
BkJlcmxpbjEPMA0GA1UEBwwGQmVybGluMQwwCgYDVQQKDANPcmcxEjAQBgNVBAMM
CWxvY2FsaG9zdDCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAJnGjlvN
O3F/Z7Lr/r+6Xp2DosnNwoPHhG2e61KnFzgZfaxbklal5ORpuV/gLIg7lrbpdZe7
WvK+16RanL6fLitis/tYVFyvz1MXqBYYrEoFGvVg9fOiis7hjpdZcpNDH9SngoAN
O0Wvv4T6LQS0cC7ZAFZjvmJ+RiZEbzRkNG5pUddZXbotE6htNfLgA5L1wIBgllrM
4DVkG0yNKmzqPNzfPTbdUgWCfjaQShHy1GP8KNEwFxM31F2wvQxsEb77o1S44Out
mlsi83tti6P7KjDk7w2j2zZO1X0xI8pflv3TBkJT1Am8vnk6rVnNO4pCpop3+kma
pDUEzBQmSQA5R1ECAwEAAaNTMFEwHQYDVR0OBBYEFDxFcFgPEsgsDixfKxB0uYGN
aJmzMB8GA1UdIwQYMBaAFDxFcFgPEsgsDixfKxB0uYGNaJmzMA8GA1UdEwEB/wQF
MAMBAf8wDQYJKoZIhvcNAQELBQADggEBAFih4lUbLlhKwIAV9x3/W7Mih8xUEdZr
olquZgaHedFet+ByAHvoES3pec7AVYTOD53mjgyZubD6INnVHzKyS4AG9ydD73o4
cmm3DKxBaesvlHeTn0MOKsoM8QCxeyFJmiUPpgDBok/PFnbGR9+JcsrlGJAnsSKD
vWpiwYcBauZ9nnK5yDe5M9XNFPkNDZzbKvWU7Sw3ziMT/+bRJse5vTrYcyOnNGgy
gZNz2nimKy1U8XZVAVwOV0rdGEFrfMln8DkRW86rGK/EncaVsl0SSP/rmjQgiX8Q
3CZraQGujJP932HSwUfdCX9yh+rTjE3MEnbqMoLzJa4BXB2aDQWtywU=
-----END CERTIFICATE-----
`

func TestParseCerts(t *testing.T) {
	certs := parseCerts(exampleCert)
	assert.Equal(t, len(certs), 1)
	assert.Equal(t, "localhost", certs[0].Subject.CommonName)
}
```
maybe?

mvo5 reviewed (commented):
Thank you for this PR! It looks good but I still have some ideas/suggestions inline for your consideration. But I'm happy to help with them if you have no enough time :)

lzap commented:
I have extracted the checking code + test into separate package `cert` and implemented an explicit check method. The error is now explicit, no warnings.

mvo5 commented on pkg/blueprint/customizations.go:
```diff
@@ -425,3 +427,27 @@ func (c *Customizations) GetRHSM() *RHSMCustomization {
+		return nil
+	}
+
+	if c.CACerts == nil {
```
This should be `!=`, no? If so ideally that would have triggered a test failure :/

lzap replied:
Correct, thanks.

mvo5 commented on pkg/manifest/os.go:
```diff
@@ -822,6 +826,25 @@ func (p *OS) serialize() osbuild.Pipeline {
 	}
 
+	if len(p.CACerts) > 0 {
+		for _, cc := range p.CACerts {
```
I would love to see a test for this but I understand this area is really difficult to test so fine to ignore that (but I also can't help to mention it)

mvo5 reviewed (commented):
Thanks for the update, looks very nice now, I made a suggestion inline, hope it's useful.

lzap commented:
Amended the recommended changes, fixed tests.

Let me know where and how to add a test for this.

REMINDER: For myself - the commit should be split as requested.

mvo5 commented on pkg/cert/parsecerts_test.go:
```diff
@@ -0,0 +1,46 @@
+	assert.Nil(t, err)
+	assert.Equal(t, len(certs), 2)
+	assert.Equal(t, "localhost", certs[0].Subject.CommonName)
+}
```
A test with a borken CERT would be super nice here (and probably not a lot of work) so that we can be sure we error on invalid cert unbles

lzap replied:
A nice one, yeah, it turns out the function silently returns an empty slice. Explicit error is better I think.

mvo5 replied:
Great to hear and thanks for adding the test!

mvo5 reviewed (commented):
This looks good now from my PoV, one tiny suggestion about a test.

lzap commented:
Rebased, hopefully it passes now.

lzap commented:
Any other comments? I would love to get this in.

achilleas-k commented on pkg/manifest/os.go:
```diff
@@ -822,6 +826,25 @@ func (p *OS) serialize() osbuild.Pipeline {
+				panic(fmt.Errorf("failed to parse CA certificates: %v", err))
+			}
+
+			for _, c := range certs {
+				path := filepath.Join("/etc/pki/ca-trust/source/anchors", filepath.Base(c.SerialNumber.Text(16))+".pem")
+				f, err := fsnode.NewFile(path, nil, "root", "root", pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: c.Raw}))
+				if err != nil {
+					panic(err)
+				}
+				pipeline.AddStages(osbuild.GenFileNodesStages([]*fsnode.File{f})...)
+			}
```
Is it worth extracting this into a helper function, just for easier reading?

achilleas-k commented on pkg/manifest/os.go:
```diff
@@ -822,6 +826,25 @@ func (p *OS) serialize() osbuild.Pipeline {
 		}
 	}
 
+	if len(p.CACerts) > 0 {
+		for _, cc := range p.CACerts {
+			certs, err := cert.ParseCerts(cc)
+			if err != nil {
+				panic(fmt.Errorf("failed to parse CA certificates: %v", err))
+			}
+
+			for _, c := range certs {
+				path := filepath.Join("/etc/pki/ca-trust/source/anchors", filepath.Base(c.SerialNumber.Text(16))+".pem")
+				f, err := fsnode.NewFile(path, nil, "root", "root", pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: c.Raw}))
+				if err != nil {
+					panic(err)
+				}
+				pipeline.AddStages(osbuild.GenFileNodesStages([]*fsnode.File{f})...)
+			}
+		}
+		pipeline.AddStage(osbuild.NewCAStageStage())
+	}
+
```
Please move this whole block up before the selinux stage is added (L798), so that any newly created files get labelled correctly.

achilleas-k reviewed (changes requested):
I don't feel too strongly about extracting the two lines into a function but the stage should be moved so that it's before the selinux one.

lzap commented:
I think the MTLS change kicked in and now breaking tests I need to fix this prior this PR:

```
internal/cloudapi/v2/server.go:231:55: cannot convert source (variable of type ostree.SourceSpec) to type worker.OSTreeResolveSpec
```

lzap commented:
For the record, tests are broken due to other change that needs to be merged first.

lzap commented:
Rebased, `images` library is now updated and it works fine.

lzap commented:
I think those failures are not relevant?!?

ezr-ondrej commented:
I don't know what the failures mean TBH :(

achilleas-k commented:
The failures are because there's a file source reference in a stage without it appearing in the manifest's sources.

Generating a manifest from this branch and inspecting it, I see the same:
```
› go run ./cmd/gen-manifests --distros rhel-9.5 --arches x86_64 --types qcow2 --config ./test/configs/all-customizations.json --output ./manifests --metadata=false
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/centos-10.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/centos-9.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/fedora-39.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/fedora-40.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/fedora-41.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-10.0.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-7.9.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-8.10.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-8.4.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-8.5.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-8.6.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-8.7.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-8.8.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-8.9.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-9.0.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-9.1.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-9.2.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-9.3.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-9.4.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-9.5.json
INFO[0000] Loaded repository configuration file: /home/achilleas/projects/osbuild/images/test/data/repositories/rhel-9.6.json
'-config' was provided, thus ignoring '-config-map' option
Collecting jobs
Collected 1 jobs
Initialised 16 workers
Submitting 1 jobs... done
Starting job rhel_9.5-x86_64-qcow2-all_customizations.json
WARN[0000] Failed to load consumer certs: no consumer key found
No match for group package "iwl5150-firmware"
No match for group package "iwl6050-firmware"
No match for group package "firewalld"
No match for group package "iwl7260-firmware"
No match for group package "dracut-config-rescue"
No match for group package "iwl105-firmware"
No match for group package "iwl2000-firmware"
No match for group package "iwl100-firmware"
No match for group package "iwl6000g2a-firmware"
No match for group package "iwl1000-firmware"
No match for group package "iwl3160-firmware"
No match for group package "iwl135-firmware"
No match for group package "iwl2030-firmware"
No match for group package "iwl5000-firmware"
Finished job rhel_9.5-x86_64-qcow2-all_customizations.json
 == Jobs == Queue:    0  Active:    0  Total:    1
RPM metadata cache kept in /tmp/rpmmd

› osbuild --inspect ./manifests/rhel_9.5-x86_64-qcow2-all_customizations.json
Traceback (most recent call last):
  File "/usr/bin/osbuild", line 8, in <module>
    sys.exit(osbuild_cli())
             ^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/osbuild/main_cli.py", line 136, in osbuild_cli
    manifest = fmt.load(desc, index)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/osbuild/formats/v2.py", line 379, in load
    load_pipeline(desc, index, manifest, source_refs)
  File "/usr/lib/python3.12/site-packages/osbuild/formats/v2.py", line 353, in load_pipeline
    load_stage(desc, index, pl, manifest, source_refs)
  File "/usr/lib/python3.12/site-packages/osbuild/formats/v2.py", line 322, in load_stage
    load_input(name, desc, index, stage, manifest, source_refs)
  File "/usr/lib/python3.12/site-packages/osbuild/formats/v2.py", line 277, in load_input
    raise ValueError(f"Unknown source reference(s) {unknown_refs}")
ValueError: Unknown source reference(s) {'sha256:4c4e8c734e4ee3a117ca8f9363ba7b706d02bfc8f297c00b02ce02babddef51c'}
```

achilleas-k commented on pkg/manifest/os.go:
```diff
@@ -791,6 +793,20 @@ func (p *OS) serialize() osbuild.Pipeline {
+				panic(err.Error())
+			}
+
+			if len(files) > 0 {
+				pipeline.AddStages(osbuild.GenFileNodesStages(files)...)
+			}
```
These files should be added to `p.Files` so that they're returned from the pipeline's inline file getter:
https://github.com/osbuild/images/blob/7f9b993f83ed0f8f4227cd5912f5be3c2f9fc7b4/pkg/manifest/os.go#L894-L903

Which is called here:
https://github.com/osbuild/images/blob/abb3328d2718431316c5c5a6674335f525b07889/pkg/manifest/manifest.go#L154-L164
so that all the needed sources are collected for the manifest generation.

lzap commented:
For the record, the missing file is the CA file that needs to be copied:

```json
{           
          "type": "org.osbuild.copy",
          "inputs": {
            "file-4c4e8c734e4ee3a117ca8f9363ba7b706d02bfc8f297c00b02ce02babddef51c": {
              "type": "org.osbuild.files",
              "origin": "org.osbuild.source",
              "references": [
                { 
                  "id": "sha256:4c4e8c734e4ee3a117ca8f9363ba7b706d02bfc8f297c00b02ce02babddef51c"
                }     
              ]     
            }     
          },    
          "options": {
            "paths": [
              {
                "from": "input://file-4c4e8c734e4ee3a117ca8f9363ba7b706d02bfc8f297c00b02ce02babddef51c/sha256:4c4e8c734e4ee3a117ca8f9363ba7b706d02bfc8f297c00b02ce02babddef51c",
                "to": "tree:///etc/pki/ca-trust/source/anchors/27894af897dd2423607045716438a725f28a6d0b.pem",
                "remove_destination": true
              } 
            ] 
          } 
        },
```

lzap commented:
It looks like it works fine:

```
go run ./cmd/gen-manifests --distros rhel-9.5 --arches x86_64 --types qcow2 --config ./test/configs/all-customizations.json --output ./manifests --metadata=false

sudo osbuild ./manifests/rhel_9.5-x86_64-qcow2-all_customizations.json --export build --export os --export image --output-directory /tmp/osbuild-test --store /tmp/osbuild-test
```

```
lzap@dev:/tmp/osbuild-test/os/etc/pki/ca-trust/source/anchors$ cat 27894af897dd2423607045716438a725f28a6d0b.pem 
-----BEGIN CERTIFICATE-----
MIIDszCCApugAwIBAgIUJ4lK+JfdJCNgcEVxZDinJfKKbQswDQYJKoZIhvcNAQEL
BQAwaDELMAkGA1UEBhMCVVMxFzAVBgNVBAgMDk5vcnRoIENhcm9saW5hMRAwDgYD
VQQHDAdSYWxlaWdoMRAwDgYDVQQKDAdSZWQgSGF0MRwwGgYDVQQDDBNUZXN0IENB
IGZvciBvc2J1aWxkMCAXDTI0MDkwMzEzMjkyMFoYDzIyOTgwNjE4MTMyOTIwWjBo
MQswCQYDVQQGEwJVUzEXMBUGA1UECAwOTm9ydGggQ2Fyb2xpbmExEDAOBgNVBAcM
B1JhbGVpZ2gxEDAOBgNVBAoMB1JlZCBIYXQxHDAaBgNVBAMME1Rlc3QgQ0EgZm9y
IG9zYnVpbGQwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDeA7OcWTrV
gstoBsUaeJKm8nelg7Lc0WNXH6yOTLsr4td4yHs0YOvFGwgSf+ffV3RAG1mgqnMG
MgkD2+z+7QhHbHHs3y0d0zfhA2bg0KVvfCWk7fNRPHY0UOePpXk245Bfw3D0VTpl
F7nePk1I7ZY09snPWUeb2rjKXzYjKjzM0h27+ykV8I8+FbdyPk/pR8whyDqtHLUa
XfFy2TFloDSYMkHKVd38BnL0bj91x5F+KsZkN4HzfbYwxLbCQfOSgy7q6TWce9kq
Lo6tya9vuvpWFm1dye7L+BodAQAq/dI/JMeCfyTb0eFb+tyzfr5aVIoqqDN+p9ft
cw4OefpHbhtNAgMBAAGjUzBRMB0GA1UdDgQWBBRV2A9YmusekPzu5Yf08cV0oPL1
wjAfBgNVHSMEGDAWgBRV2A9YmusekPzu5Yf08cV0oPL1wjAPBgNVHRMBAf8EBTAD
AQH/MA0GCSqGSIb3DQEBCwUAA4IBAQCgQZ2Xfj+NxaKBZgn2KNxS0MTbhzHRz6Rn
qJs+h8OUz2Crmaf6N+RHlmDRZXUrDjSHpxVT2LxFy7ofRrLYIezFDUYfb920VkkV
SVcxh1YDFROJalfMoE6wdyR/LnK4MJZS9fUpeCJJc/A0J+9FK9CwcyUrHgJ8XbJh
MKYyQ+cf6O7wzutuBpMyRqSKS+hVM7BQTmSFvv1eAJlo6klGAmmKiYmAEvcQadH1
djrujsA3Cn5vX2L+0yuiLB5/zoxqx5cEy97TuKUYB8OqMMujAXNzF4L3HJDUNba2
AhEkFozMXwYX73TGbGZ0mawPS5D3v3tYTEmJFf6SnVCmUW1fs57g
-----END CERTIFICATE-----
```

mvo5 commented on pkg/osbuild/pki_update_ca_trust_stage_test.go:
```diff
@@ -0,0 +1,67 @@
+	assert.Nil(t, err)
+	assert.Equal(t, 2, len(certs), len(files))
+	for i, c := range certs {
+		assert.Equal(t, pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: c.Raw}), files[i].Data())
```
(nitpick) it might be nice to double check that the path setting works correctly:
```go
diff --git a/pkg/osbuild/pki_update_ca_trust_stage_test.go b/pkg/osbuild/pki_update_ca_trust_stage_test.go
index 8e20cde9e..b07a60fff 100644
--- a/pkg/osbuild/pki_update_ca_trust_stage_test.go
+++ b/pkg/osbuild/pki_update_ca_trust_stage_test.go
@@ -61,7 +61,12 @@ func TestNewCAFileNodes(t *testing.T) {
        files, err := NewCAFileNodes(exampleCert)
        assert.Nil(t, err)
        assert.Equal(t, 2, len(certs), len(files))
+       expectedPaths := []string{
+               "/etc/pki/ca-trust/source/anchors/5726bb549dcef16f12ab0b846b4059e074ac5ef0.pem",
+               "/etc/pki/ca-trust/source/anchors/4550f533dccaa359ad1b20d2ddbd27cd0d0de24.pem",
+       }
        for i, c := range certs {
+               assert.Equal(t, expectedPaths[i], files[i].Path())
                assert.Equal(t, pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: c.Raw}), files[i].Data())
        }
 }

``` 
(or something like it)

mvo5 commented on pkg/manifest/os.go:
```diff
@@ -791,6 +793,21 @@ func (p *OS) serialize() osbuild.Pipeline {
+		for _, cc := range p.CACerts {
+			files, err := osbuild.NewCAFileNodes(cc)
+			if err != nil {
+				panic(err.Error())
```
(super nitpick) `panic(err)` would also work (and is slightly shorter)

lzap replied:
This looks like a pattern in the whole file, I would rather not change it in this PR.

mvo5 commented on pkg/cert/parsecerts.go:
```diff
@@ -0,0 +1,44 @@
+
+// ParseCerts parses a PEM-encoded certificate chain formatted as concatenated strings
+// and returns a slice of x509.Certificate. In case of unparsable certificates, the
+// function returns an empty slice.
```
(niptick) should this also say that an error is returned? I.e. `In case of unparsable certificates, the
// function returns an empty slice and an error`?

mvo5 commented on pkg/distro/fedora/images.go:
```diff
@@ -226,6 +226,14 @@ func osCustomizations(
 
+	ca, err := c.GetCACerts()
+	if err != nil {
+		panic(fmt.Sprintf("unexpected error checking CA certs: %v", err))
```
Hm, I am not sure we should panic here, I see we do this a lot in this function whichI find confusing given that we have an error return here and that the certs come from the user so they maybe wrong and panic() seems a bit heavy handed for user inputs. But we can tweak in a followup given the rest of the function.

lzap replied:
Ditto.

mvo5 commented on pkg/distro/rhel/images.go:
```diff
@@ -271,6 +271,14 @@ func osCustomizations(
 
+	ca, err := c.GetCACerts()
+	if err != nil {
+		panic(fmt.Sprintf("unexpected error checking CA certs: %v", err))
```
Same comment as above, panic() seems heavy handed here

mvo5 reviewed (dismissed):
Thank you! This looks fine, I have some inline suggestions but no blockers, this seems all fine now. Thanks also for splitting it up.

lzap commented:
Rebased.

mvo5 reviewed (approved):
Thank you!

lzap commented:
Forgot to run `prepare-source.sh` fixed, rebased.
